### PR TITLE
Adds ASP.NET Document

### PR DESCRIPTION
**Pull request by andrewabest**
Adds a document intended to give an overview of our ASP.NET web framework implementation, and provide some guidance to developers looking to develop new ASP.NET endpoints within Octopus Server

### CODE REVIEW

**Comment by jbristowe** [comment]
👍

File: OctopusServer/Api/AspNet.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+To help developers remember the parts they need to put in place for each Controller, we have established a suite of [Controller Conventions](TODO) that inspect all Controllers created and will not pass unless the required structures and attributes are in place.
+
+# Developing
+
+## Creating New ASP.NET Controllers
+
+Creating a new ASP.NET Controller should be straight-forward.
+
+Add a new Controller class under the appropriate namespace, derive from `SpaceScopedApiController` to ensure your request is established in an appropriate Space Partition, and ensure you add the required attributes to the Controller endpoint and any incoming parameters. Running the [Controller Conventions](TODO) will help you discover anything you may have missed.
```
Comment (paulegradie) [inline OctopusServer/Api/AspNet.md:65]: Not all controllers will be space scoped

File: OctopusServer/Api/AspNet.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+- Performance of new ASP.NET Controllers was on average ~25% better (note that this is baseline performance on stable hardware, not under load / variable conditions)
+- With correct attribution (enforced by convention tests), Swagger output was improved
+- Wire compatibility of responses was as-expected, with some improvement on some existing Nancy-isms (particularly around the handling of chunked encoding)
+
+For future migrations, implementors may want to consider the risk / complexity involved in their particular migration, and consider:
+
+- Existing E2E test coverage that exercises the endpoint(s) in question
+- Whether additional coverage utilizing [AssentTestScript] within the [Performance Test Suite](https://github.com/OctopusDeploy/OctopusDeploy/blob/master/source/Octopus.E2ETests.Performance/Readme.md) would provide additional confidence
+
+We don't think additional performance testing for regression is required given our observed results during the Tenants migration, unless there is a particular reason to suspect performance might change as a result of the migratino.
```
Comment (paulegradie) [inline OctopusServer/Api/AspNet.md:86]: migration

File: OctopusServer/Api/AspNet.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+### Legacy Responders (Legacy Nancy Request Processing)
+
+Will use one of the Responder registration methods declared on `OctopusNancyModule` (like `Load`, `Create`, or `Modify`), with responders implementing [Responder](https://github.com/OctopusDeploy/OctopusDeploy/blob/master/source/Octopus.Server/Web/Infrastructure/Api/Responder.cs), and may use `PersistenceRule`s to enforce validation or create other side-effects.
```
Comment (paulegradie) [inline OctopusServer/Api/AspNet.md:90]: I didn't automatically prefix 'legacy reponders' to this section, and had to reread it a couple times XD

File: OctopusServer/Api/AspNet.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+To migrate legacy responders, all `PersistenceRule` code should be re-implemented within the Controller or domain model, and the existing rules deleted. It is important to consider where in the Responder lifecycle the rules were applied, so that the behaviour can be accurately replicated within the Controller.
+
+See the [ModifyTenantController](https://github.com/OctopusDeploy/OctopusDeploy/blob/master/source/Octopus.Server/Web/Controllers/Tenants/ModifyTenantController.cs)
+
+### Custom Responders (Newer Nancy Request Processing)
+
+Will use the `CustomAction` registration method on `OctopusNancyModule` and implement `Custom{Action/Create/Modify}Responder`. Rules utilized will implement `IResponderStep`, with a simple 'veto' pattern used for rule enforcement directly from the responder.
+
+To migrate Custom Responders, the default recommendation is taking the `IResponderStep` rules as dependencies to the Controller, and exercising them within the new Controller created.
```
Comment (Waldo000000) [inline OctopusServer/Api/AspNet.md:100]: Suggest adding a link to an PR that shows an example of such a migration.

File: OctopusServer/Api/AspNet.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+We don't think additional performance testing for regression is required given our observed results during the Tenants migration, unless there is a particular reason to suspect performance might change as a result of the migratino.
+
+### Legacy Responders (Legacy Nancy Request Processing)
+
+Will use one of the Responder registration methods declared on `OctopusNancyModule` (like `Load`, `Create`, or `Modify`), with responders implementing [Responder](https://github.com/OctopusDeploy/OctopusDeploy/blob/master/source/Octopus.Server/Web/Infrastructure/Api/Responder.cs), and may use `PersistenceRule`s to enforce validation or create other side-effects.
+
+To migrate legacy responders, all `PersistenceRule` code should be re-implemented within the Controller or domain model, and the existing rules deleted. It is important to consider where in the Responder lifecycle the rules were applied, so that the behaviour can be accurately replicated within the Controller.
+
+See the [ModifyTenantController](https://github.com/OctopusDeploy/OctopusDeploy/blob/master/source/Octopus.Server/Web/Controllers/Tenants/ModifyTenantController.cs)
```
Comment (Waldo000000) [inline OctopusServer/Api/AspNet.md:94]: Suggest adding a link to an PR that shows an example of such a migration.

File: OctopusServer/Api/AspNet.md (hunk 4 above)
Comment (andrewabest) [reply OctopusServer/Api/AspNet.md:100]: We don't have one of these yet - but I'll link it if we do get one.

File: OctopusServer/Api/AspNet.md (hunk 1 above)
Comment (andrewabest) [reply OctopusServer/Api/AspNet.md]: Updated

File: OctopusServer/Api/AspNet.md (hunk 2 above)
Comment (andrewabest) [reply OctopusServer/Api/AspNet.md]: Done

File: OctopusServer/Api/AspNet.md (hunk 5 above)
Comment (andrewabest) [reply OctopusServer/Api/AspNet.md]: Done

**Pull request merged**